### PR TITLE
Enhance finding of qhull

### PR DESCRIPTION
**Pull request by larshg**
closes #3741 and #3740.

### CODE REVIEW

**Comment by elfprince13** [comment]
Incidentally, PCL 1.12.0 is broken on Ubuntu 20.04 with `libqhull-dev` installed from apt, and this patch looks like it will fix things (as `qh_zero` has been moved to `libqhull_r.so` - `libqhull.so` is now a stub).

**Comment by mvieth** [comment]
> Incidentally, PCL 1.12.0 is broken on Ubuntu 20.04 with `libqhull-dev` installed from apt, and this patch looks like it will fix things (as `qh_zero` has been moved to `libqhull_r.so` - `libqhull.so` is now a stub).

Could you explain a bit more in which way it is broken? One of our CI jobs runs Ubuntu 20.04 with `libqhull-dev` installed via apt, and that builds fine (even without this pull request)

**Comment by elfprince13** [comment]
Linking against the targets created by `FindPCL.cmake` yields a bunch of qhull-related linking errors, because it tries to link against the empty libqhull.so rather than the populated libqhull_r.so.

**Comment by elfprince13** [comment]
Sorry - I meant - `PCLConfig.cmake`.

Compilation and install sequence is:

```bash
mkdir -p build/usr
cd build
cmake -DCMAKE_INSTALL_PREFIX:PATH=$(pwd)/usr -DQHULL_USE_STATIC=FALSE -DBUILD_visualization=OFF -DWITH_PCAP=FALSE -DWITH_OPENGL=FALSE -DWITH_VTK=FALSE ..
make
make install
```

Then in the client project:
```cmake
set(PCL_DIR /path/to/pcl/build/usr/share/pcl-1.12)
find_package(PCL 1.12.0 REQUIRED)
```

**Comment by larshg** [comment]
Anyone have an idea why the MacOS builds are failing due to missing <omp.h> in eigen include?

**Comment by mvieth** [comment]
> Anyone have an idea why the MacOS builds are failing due to missing <omp.h> in eigen include?

Maybe the directory where omp.h can be found, was part of `QHULL_INCLUDE_DIRS`? Maybe linking the vector average test with `pcl_common` solves that? (all the other tests in common do that)

**Comment by larshg** [comment]
> > Anyone have an idea why the MacOS builds are failing due to missing <omp.h> in eigen include?
> 
> Maybe the directory where omp.h can be found, was part of `QHULL_INCLUDE_DIRS`? Maybe linking the vector average test with `pcl_common` solves that? (all the other tests in common do that)

Thanks @mvieth - seems you were right. Apparently there was some more that needed to be added.

**Comment by larshg** [comment]
@SunBlack would you review this and maybe run it through your CI 😄 ?

**Comment by SunBlack** [comment]
Is this MR on the current master? As the CI errors looks like not (so either you rebase this or I need to add a merge step in our CI ;-) )

**Comment by larshg** [comment]
> Is this MR on the current master? As the CI errors looks like not (so either you rebase this or I need to add a merge step in our CI ;-) )

Ahh right, I'll rebase asap.

**Comment by larshg** [comment]
> The CI is green here :-).
> 
> Didn't had now time to make a full review (didn't know we have already sth. like `DONTCARE`). But one thing: Can you parse the Qhull version from any file and pass the value to CMake (so we can set a required minimum version if necessary)?

I have looked, but cannot find any file with the version, except for qhull 8 which has the Cmake config files. So it seems not possible for now.

File: cmake/Modules/FindQhull.cmake [hunk 1]
```diff
@@ -1,24 +1,107 @@
...
-if(QHULL_USE_STATIC)
-  set(QHULL_RELEASE_NAME qhullstatic_r)
-  set(QHULL_DEBUG_NAME qhullstatic_rd)
-else()
-  set(QHULL_RELEASE_NAME qhull_r qhull)
-  set(QHULL_DEBUG_NAME qhull_rd qhull_d)
+# Try to find QHULL library and headers. This module supports both old released versions
+# of QHULL ≤ 7.3.2 and newer development versions that ship with a modern config file.
+#
+# If QHULL_USE_STATIC is specified then look for static libraries ONLY else look for shared one.
```
[the remark this answers is not in the document]
Comment (larshg) [reply cmake/Modules/FindQhull.cmake]: Updated the documentation.

**Comment by SunBlack** [comment]
> I have looked, but cannot find any file with the version, except for qhull 8 which has the Cmake config files. So it seems not possible for now.

Downloaded a old deb file and checked the content. I see two variants:
- Some headers habe a line `$$Change: 2953 $` which seems to be adjusted which each minor release (but not the patch version), so be holding a list it would be possible.
- Use `try_compile` to extract value of `qh_version2[]` or `qh_version[]`

Benefit of this we see in the CMake log which version is used in case there is an issue.

**Comment by larshg** [comment]
> > I have looked, but cannot find any file with the version, except for qhull 8 which has the Cmake config files. So it seems not possible for now.
> 
> Downloaded a old deb file and checked the content. I see two variants:
> 
>     * Some headers habe a line `$$Change: 2953 $` which seems to be adjusted which each minor release (but not the patch version), so be holding a list it would be possible.
> 
>     * Use `try_compile` to extract value of `qh_version2[]` or `qh_version[]`
> 
> 
> Benefit of this we see in the CMake log which version is used in case there is an issue.

Seems like a lot of effort for something that is mostly needed for all previous versions of ubuntu before 20.10, since here its shown that its version 8.
I'll leave it up as a followup PR, since I see only minor benefit from exactly knowing the version.

**Review by mvieth — DISMISSED** [review]
Looks good to me.
Do you want to clean up the commit history by hand or just squash everything when merging?
I think we can add parsing and displaying the version in the future, but would also not do this here, in the interest of merging this soon so we can release PCL 1.12.1

**Review by kunaltyagi — DISMISSED** [review]
❤️

**Comment by mvieth** [comment]
The else-case you added (FindQhull.cmake starting at line 84) is exactly the `DONTCARE` case, right? What do you think about removing the `DONTCARE` case then and let it fall through to the final else? (And maybe add a comment that it covers `DONTCARE`)

**Comment by larshg** [comment]
The else case is if you have selected ie. static, but its not found but can find the shared version it will tell that static was selected but not found, but shared was found.

The first if, is if you don't care yes and then it selects according to PCL_SHARED_BUILD and what is available. But I can see I missed a PCL_SHARED_BUILD AND in the else if with the static version.

**Comment by larshg** [comment]
Sooo, I looked at how I implemented it with FLANN library and realised it could be done in the same way. So its rewritten again 😄

**Review by kunaltyagi — DISMISSED** [review]
Would it be possible to make a macro if there's some sharable part?

File: cmake/Modules/FindQhull.cmake [hunk 2]
```diff
@@ -1,90 +1,194 @@
...
-# look for shared ones
 
-if(QHULL_USE_STATIC)
-  set(QHULL_RELEASE_NAME qhullstatic_r)
-  set(QHULL_DEBUG_NAME qhullstatic_rd)
-else()
-  set(QHULL_RELEASE_NAME qhull_r qhull)
-  set(QHULL_DEBUG_NAME qhull_rd qhull_d)
+
+macro(ExtractProperties target)
```
Comment (larshg) [inline cmake/Modules/FindQhull.cmake:40]: Meh, forgot to remove this one.

File: cmake/Modules/FindQhull.cmake [hunk 3]
```diff
@@ -1,90 +1,194 @@
...
+        set(QHULL_LIBRARY_TYPE STATIC)
+      endif()
+    elseif(PCL_QHULL_REQUIRED_TYPE MATCHES "SHARED")
+      set_property(TARGET QHULL::QHULL APPEND PROPERTY INTERFACE_LINK_LIBRARIES Qhull::qhull_r)
+      set(QHULL_LIBRARY_TYPE SHARED)
+    else()
+      set_property(TARGET QHULL::QHULL APPEND PROPERTY INTERFACE_LINK_LIBRARIES Qhull::qhullstatic_r)
+      set(QHULL_LIBRARY_TYPE STATIC)
+    endif()
+  elseif(TARGET qhull::qhullstatic_r)
```
Comment (mvieth) [inline cmake/Modules/FindQhull.cmake:85]: Does capitalization matter? `qhull::` vs `Qhull::`?

Comment (larshg) [reply cmake/Modules/FindQhull.cmake:73]: Yes. Thanks I'll fix it.

**Comment by larshg** [comment]
> Would it be possible to make a macro if there's some sharable part?

I don't think it would be possible, at least not easy.

**Review by mvieth — APPROVED** [review]

**Review by kunaltyagi — APPROVED** [review]

**Comment by kunaltyagi** [comment]
Looking at commit history, it might make sense to squash merge this

**Pull request merged**